src/Model/Operation/ProductSyncHandler.php

```diff
@@ -178,35 +201,55 @@ private function doUpsertOperation(
         $operation->upsert();
     }
 
-    private function processProductVariants(SalesChannelProductEntity $product, SalesChannelContext $context): array
-    {
+    private function processProductVariants(
+        ProductEntity $product,
+        SalesChannelContext $context,
+        Account $account,
+        array $ids,
+    ): ProductCollection {
         $variantConfig = $product->getVariantListingConfig();
         $configuratorGroups = array_filter(
             $variantConfig?->getConfiguratorGroupConfig() ?? [],
             static fn (array $config) => $config['expressionForListings'],
         );
 
         if (!$product->getChildCount() || !($variantConfig instanceof VariantListingConfig)) {
-            return [$product];
+            return new ProductCollection([$product]);
         }
 
         $mainProducts = [];
         if ($variantConfig->getDisplayCheapestVariant()) {
             $mainProducts[] = $this->handleCheapestVariant($product, $context);
         } elseif ($variantConfig->getMainVariantId()) {
-            $mainProducts[] = $this->handleMainVariant($product, $variantConfig);
+            if ($variant = $this->handleMainVariant($product, $variantConfig)) {
+                $mainProducts[] = $variant;
+            }
         } elseif (count($configuratorGroups)) {
             $mainProducts = array_merge(
                 $mainProducts,
                 $this->handleConfiguratorGroups($product),
             );
+        } elseif (!$variantConfig->getDisplayParent() || !$product->getActive()) {
+            if ($variant = $this->handleFirstActiveVariant($product)) {
+                $mainProducts[] = $variant;
+            }
         }
 
-        return count($mainProducts) ? $mainProducts : [$product];
+        if (!$mainProducts) {
```

Comment:
After my suggested change from above:

```suggestion
        if (!$mainProducts->count()) {
```